kizhin.evgeniy/S1

sdjklfjsdlkfjsdklfj opened the pull request:
Fwd | Fake
9

Commit:
Introduce SFINAE for input iterator validation

Added enable_if_input_iterator template alias to validate if a
given type is an input iterator. This will be used in the list
implementation

Commit:
Add declarations: List, iterators, comparison, swap

- Declared the List class with all member methods and constructors
- Defined List methods with stubs and TODOs
- Internal Node structure is also defined.
- Added list comparison operators, forward declaration, and swap
- Iterators implemented as template class with bool parameter that
  defines is iterator const or not

Commit:
Add boost test module for S1 and tests

- Covered constructors, assignment operator, push/pop/emplace, assign,
  insert
- Added stubs for other tests

Commit:
List to ForwardList

Commit:
Update ForwardList implementation

- Replaced default constructor with explicit initialization of begin_ and end_
- Updated constructors to use push_back instead of emplace_back
- Fixed pop_back|front to handle empty list
- Implemented emplace[_back|front], assign, insert
- Added assertions
- Refactoring
- Added size_ field to provide size() with O(1)
- Refactored existing methods to deal with size_ field

Commit:
Refactor ForwardList test suite

- Replaced std::list with kizhin::ForwardList
- Updated method names according CG
- Refactored existed tests according new list interface
- Improved test coverage for edge cases
- Marked incomplete tests with TODOs
- Divided tests with test suits
- Moved testInvariants to test-utils file

Commit:
Implement comparison operators

Commit:
Refactor ForwardList

- Removed non-const erase overload
- Removed insert(const&) and (&&) via passing by value. The same with pushBack|Front
- Implemented removeIf
- Refactored assign (via construct and swap)
- Defined splice with TODO stubs

Commit:
Implement modifiers with tests: unique, reverse, ....

- Introduced testComparisonInvariants helper function in test-utils
- Implemented tests for: remove, reverse, splice, unique
- Implemented comparison operators tests
- Implemented unique, reverse, splice

Commit:
Introduce io-utils, type-utils, list-utils

Commit:
Introduce main

Commit:
Introduce fake node before the beginning

Added beforeBegin() method to get iterator before begin()
Supported beforeBegin() in tests

Commit:
Implement merge and sort. Refactor spliceAfter

Refactored spliceAfter: strong exception safety via nodes relinking
Implemented merge & sort in ForwardList
Implemented tests for merge & sort
Added new test for spliceAfter with corner case
Cleaned up according CG as well

Commit:
Move forward-list to common

Commit:
Refactor according CG

- Updated header guards in common (S1 -> common)
- Marked ForwardList, Node and Iterator as final
- Used auto for begin() calls and other common cases
- Refactored loops with range-based for where its possible
- Introduced outputListWithNewLine function to avoid code duplicating
- Refactored lambdas in tests
- Renamed testInvariants to testForwardListInvariants
- Renamed convert to transpose
- Refactor transpose to use std::move and be more efficient
- Removed extra headers
- Added noexcept in ForwardList front,back,reverse
- Minor refactoring

Commit:
Add algorithm header to list-comparison

Commit:
Refactor: avoid line breaks

Commit:
Refactor iterators

- Renamed template arguments with CamelCase (according CG)
- Refactored SFINAE: used enable_if_t instead of enable_if
- Added assert in iterator methods
- Added noexcept to certain iterator methods and comparison operators
- Removed extra empty lines

Commit:
Add messages in assertions

Commit:
Refactor loops. Introduce algorithms header

- Introduced algorithms header with lexicographicalCompare
- Refactored loops in main
- Refactored braces split
- Refactored ForwardList constructors with insertAfter instead of loops
- Minor refactoring

Volgarenok commented on kizhin.evgeniy/S1/io-utils.hpp:
```diff
@@ -0,0 +1,41 @@
+    out << '\n';
+  }
+  return out;
+}
```
Ну это явно лишнее. Лучше это в коде видеть

Volgarenok commented on kizhin.evgeniy/S1/list-utils.cpp:
```diff
@@ -0,0 +1,43 @@
+{
+  using NumberIter = NumbersT::const_iterator;
+  ForwardList< std::pair< NumberIter, NumberIter > > iters;
+  for (const auto& nums : source) {
```
CG

Volgarenok commented on kizhin.evgeniy/S1/type-utils.hpp:
```diff
@@ -0,0 +1,14 @@
+#include <iosfwd>
+#include <forward-list-fwd-declaration.hpp>
+
+namespace kizhin {
+  using NumbersT = ForwardList< std::uint64_t >;
+  using NamesT = ForwardList< std::string >;
+}
```
Нехорошо, потому что весь код шаблонный, который вы вокруг этого написали теперь привязан к этим определениям. И получается не очень.. Что толку в этих шаблонах, если по факту вы из них практически ничего не генерируете?

Volgarenok replied:
Я к тому, что может шаблоны не нужны в этой части? Ну то есть достаточно шаблона списка как такового. А логика программы не должна быть шаблонной?

Volgarenok commented on kizhin.evgeniy/S1/main.cpp:
```diff
@@ -0,0 +1,36 @@
+    {
+      return list.empty();
+    };
+    if (names.empty() || std::all_of(numbers.begin(), numbers.end(), isEmptyList)) {
```
Передавайте указатель на метод. Без лямбды

Volgarenok replied:
Да, тут не получится. Сори, мой косяк

Volgarenok replied:
Вы тогда вынесите функтор над main-ом и нормально его оформите. Назовите как-то

Volgarenok commented on kizhin.evgeniy/common/algorithms.hpp:
```diff
@@ -0,0 +1,22 @@
+
+namespace kizhin {
+  template < typename InputIt >
+  bool lexicographicalCompare(InputIt first1, InputIt last1, InputIt first2,
```
Не хватает параметра - компаратора

Volgarenok commented on kizhin.evgeniy/common/forward-list-comparison.hpp:
```diff
@@ -0,0 +1,47 @@
+#define SPBSPU_LABS_2025_AADS_A_KIZHIN_EVGENIY_COMMON_FORWARD_LIST_COMPARISON_HPP
+
+#include <algorithm>
+#include "algorithms.hpp"
```
Ну вот очень аккуратно надо быть, когда вы такие заголовки делаете

Volgarenok commented on kizhin.evgeniy/common/forward-list-fwd-declaration.hpp:
fwd достаточно

Commit:
Improve code structure: split pushBack and replace lambda

- Added pushBack(const T&) and pushBack(T&&) instead of pushBack(T)
- Replaced lambda in main with functor

Commit:
Refactor naming: file names

In common directory:
- forward-list-fwd-declaration.hpp -> forward-list-fwd.hpp
- algorithms.hpp                   -> algorithm-utils.hpp

Commit:
Refactor: improve naming & syntax, add comparator

- Replaced outputListWithNewLine calls with outputList for clarity
- Removed space before range based for loop colon according CG
- Added comparator as parameter in lexicographicalCompare function

Commit:
Move functor from main to list-utils

Commit:
Merge master of Volgarenok/spbspu-labs-2025-aads-a into kizhin.evgeniy/S1
